### PR TITLE
Optimize reversed(range(...)), with non-constant values.

### PR DESCRIPTION
**Pull request by BigFav**
One potential problem is that is does not generate warnings on the division, however we know that `step_value != 0`. I have a much larger test file on my computer, if there are doubts of the basic functionality. Hopefully it fits into the structure properly this time, if not comments are welcome.

Completing ticket #765.

### CODE REVIEW

**Comment by BigFav** [comment]
Hey @scoder, have had the time to look at this? I just want to know if anything is wrong, or if you want me to re-pull with cleaner commits (all in one, or spread into two where one has the code changes, and the other has the tests?). Hope you had a happy new years day!

File: Cython/Compiler/Optimize.py [hunk 1]
```diff
@@ -687,7 +747,10 @@ def _transform_range_iteration(self, node, range_function, reversed=False):
         if not bound2.is_literal:
             # stop bound must be immutable => keep it in a temp var
             bound2_is_temp = True
-            bound2 = UtilNodes.LetRefNode(bound2)
+            if bound2_ref_node:
+                bound2 = bound2_ref_node
+            else:
+                bound2 = UtilNodes.LetRefNode(bound2)
```
Comment (BigFav) [inline Cython/Compiler/Optimize.py]: JavaScript's style would have this if-else replaced by `bound2 = bound2_ref_node or UtilNodes.LetRefNode(bound2)`. What's your take on this?

Comment (scoder) [inline Cython/Compiler/Optimize.py]: yes, it's ok to use that.

**Comment by scoder** [comment]
Looks good to me now.

**Comment by scoder** [comment]
Thanks for working on this!

**Comment by BigFav** [comment]
I'm glad to be able to contribute.

**Pull request merged**